File: spC++/src/Implement/ImageProxy.cpp
#include "../Headers/ImageProxy.h"

ImageProxy::ImageProxy(std::string url) {
    this->url = url;
    image = nullptr;
}

ImageProxy::ImageProxy(ImageProxy* imageProxy) {
    this->url = imageProxy->url;
    image = nullptr;
}

void ImageProxy::print() {
    if(image == nullptr) {
        image = new Image(url);
    }
    image->print();
}

IElement* ImageProxy::clone() {
    return new ImageProxy(this);
}

std::string ImageProxy::getUrl() {
    return url;
}

ImageProxy::~ImageProxy() {
    delete image;
}

Image ImageProxy::loadImage() {
    if (image == nullptr) {
        image = new Image(url);
    }
    return *image;
}

File: Java-Labs/lab/src/main/java/sp/lab/Models/Image/ImageProxy.java
package sp.lab.Models.Image;

import sp.lab.Models.Element;

public class ImageProxy extends Element implements IPicture{
    private String url;
    private Image realImage;
    private Size size;

    public ImageProxy(String url, int width, int height) {
        this.url = url;
        this.size = new Size(width, height);
        this.realImage = null;
    }

    public Image loadImage() {
        if(realImage == null) {
            realImage = new Image(this.url);
        }
        return realImage;
    }

    public void print() {
        this.loadImage();
        this.realImage.print();
    }
}
